src/providers/hana/qgshanaprovider.cpp

```diff
@@ -0,0 +1,1798 @@
+{
+  QString sql = buildQuery( QStringLiteral( "COUNT(*)" ), whereClause );
+  QgsHanaConnectionRef conn( mUri );
+  size_t count = conn->executeCountQuery( sql );
+  return static_cast<long>( count );
+}
+
+QgsCoordinateReferenceSystem QgsHanaProvider::crs() const
+{
+  QgsCoordinateReferenceSystem srs;
+
+  static QMutex sMutex;
+  QMutexLocker locker( &sMutex );
+  static QMap<int, QgsCoordinateReferenceSystem> sCrsCache;
+  if ( sCrsCache.contains( mSrid ) )
+  {
+    srs = sCrsCache.value( mSrid );
+    return srs;
+  }
+
+  QgsHanaConnectionRef conn( mUri );
+  srs = conn->getCrs( mSrid );
+
+  if ( srs.isValid() )
+    sCrsCache.insert( mSrid, srs );
+
+  return srs;
+}
+
+QgsVectorLayerExporter::ExportError QgsHanaProvider::createEmptyLayer(
+  const QString &uri,
+  const QgsFields &fields,
+  QgsWkbTypes::Type wkbType,
+  const QgsCoordinateReferenceSystem &srs,
+  bool overwrite,
+  QMap<int, int> *oldToNewAttrIdxMap,
+  QString *errorMessage,
+  const QMap<QString, QVariant> *
+)
+{
+  QgsDataSourceUri dsUri( uri );
+  QgsHanaConnectionRef conn( dsUri );
+  if ( conn.isNull() )
+  {
+    if ( errorMessage )
+      *errorMessage = QObject::tr( "Connection to database failed" );
+    return QgsVectorLayerExporter::ErrConnectionFailed;
+  }
+
+  QString schemaName = dsUri.schema();
+  QString tableName = dsUri.table();
+
+  if ( schemaName.isEmpty() )
+  {
+    if ( errorMessage )
+      *errorMessage = QObject::tr( "Schema name cannot be empty" );
+    return QgsVectorLayerExporter::ErrCreateLayer;
+  }
+
+  QString geometryColumn = dsUri.geometryColumn();
+  QString geometryType;
+
+  QString primaryKey = dsUri.keyColumn();
+  QString primaryKeyType;
+
+  QString schemaTableName = QgsHanaUtils::quotedIdentifier( schemaName ) + '.' +
+                            QgsHanaUtils::quotedIdentifier( tableName );
+
+  bool fieldsInUpperCase = false;
+  if ( fields.size() > 0 )
+  {
+    int count = QgsHanaUtils::countFieldsWithFirstLetterInUppercase( fields );
+    fieldsInUpperCase = count > fields.size() / 2;
+  }
+
+  if ( wkbType != QgsWkbTypes::NoGeometry && geometryColumn.isEmpty() )
+    geometryColumn = fieldsInUpperCase ? "GEOM" : "geom";
+
+  bool createdNewPk = false;
+
+  if ( primaryKey.isEmpty() )
+  {
+    QString pk = primaryKey = fieldsInUpperCase ? "ID" : "id";
+    int index = 0;
+    while ( fields.indexFromName( primaryKey ) >= 0 )
+    {
+      primaryKey = QStringLiteral( "%1_%2" ).arg( pk ).arg( index++ );
+    }
+
+    createdNewPk = true;
+  }
+  else
+  {
+    int idx = fields.indexFromName( primaryKey );
+    if ( idx >= 0 )
+    {
+      QgsField fld = fields.at( idx );
+      if ( QgsHanaUtils::convertField( fld ) )
+        primaryKeyType = fld.typeName();
+    }
+  }
+
+  if ( primaryKeyType.isEmpty() )
+    primaryKeyType = QStringLiteral( "BIGINT" );
+
+  QString sql;
+
+  // set up spatial reference id
+  long srid = 0;
+  if ( srs.isValid() )
+  {
+    srid = srs.postgisSrid();
```

Comment:
This shouldn't be used -- it's a deprecated method which will be removed, and isn't safe even for the postgis provider

Reply:
Which one should we use instead?

Reply:
We shouldn't use any numeric IDs on the QGIS side -- everything should ideally be done using auth:id matches or WKT strings.

Reply:
Done in 1fa8b839a6.